File: CodeSignal/javascript/sudoku2.js
// Sudoku is a number-placement puzzle. The objective is to fill a 9 × 9 grid with numbers in such a way that each column, each row, 
// and each of the nine 3 × 3 sub-grids that compose the grid all contain all of the numbers from 1 to 9 one time.

// Implement an algorithm that will check whether the given grid of numbers represents a valid Sudoku puzzle according to the layout rules described above. 
// Note that the puzzle represented by grid does not have to be solvable.

// U
// basic rules of sudoku
// no number should appear on same row/col more than once
// implement an algo to check if numbers on grid rep a valid puzzle
// if valid return true else return false

// P
// traverse through grid
// check if row/col have unique integers 
// if valid return true else return false


sudoku = [['.', '.', '.', '1', '4', '.', '.', '2', '.'],
          ['.', '.', '6', '.', '.', '.', '.', '.', '.'],
          ['.', '.', '.', '.', '.', '.', '.', '.', '.'],
          ['.', '.', '1', '.', '.', '.', '.', '.', '.'],
          ['.', '6', '7', '.', '.', '.', '.', '.', '9'],
          ['.', '.', '.', '.', '.', '.', '8', '1', '.'],
          ['.', '3', '.', '.', '.', '.', '.', '.', '6'],
          ['.', '.', '.', '.', '.', '7', '.', '.', '.'],
          ['.', '.', '.', '5', '.', '.', '.', '7', '.']]

// E
function sudoku2(grid) {
    for(var row = 0; row < 9; row++){
        for(var col = 0; col < 9; col++){
            if(grid[row][col] == grid[col][row]){
                return true
            } else {
                return false
            }
        }
    }
}

console.log(sudoku2(sudoku))

// R
// 

File: CodeSignal/python/arrays/sudoku2.py
#           line[i]
#              v
g =    [['.', '.', '.', '1', '4', '.', '.', '2', '.'],
        ['.', '.', '6', '.', '.', '.', '.', '.', '.'],
        ['.', '.', '.', '.', '.', '.', '.', '.', '.'],
        ['.', '.', '1', '.', '.', '.', '.', '.', '.'],
        ['.', '6', '7', '.', '.', '.', '.', '.', '9'],
        ['.', '.', '.', '.', '.', '.', '8', '1', '.'],
        ['.', '3', '.', '.', '.', '.', '.', '.', '6'],
        ['.', '.', '.', '.', '.', '7', '.', '.', '.'],
        ['.', '.', '.', '5', '.', '.', '.', '7', '.']]


def check_unique(nums):
    # create empty set to add num values in
    s = set()
    # loop through each item in line 
    for num in nums:
        # if box is 'empty' continue
        if num == '.':
            continue
        # if num is in set already 
        if num in s:
            # it is not unique
            return False
        # otherwise add num to set
        s.add(num)
    # return True if each num is unique in line
    return True

def sudoku2(grid):
    # loop thru each row in grid
    for line in grid:
        # check if each number in row is unique 
        if not check_unique(line):
            return False
    # loop through each item in col
    for i in range(9):
        # line[i] is each column of line in grid
        if not check_unique([line[i] for line in grid]):
            return False
    # loop through each sudoku box (3x3) to see if unique
    # for row of 3
    for i in range(0,9, 3):
        # for col of 3
        for j in range(0, 9, 3):
            # check if each box has unique integers
            if not check_unique(grid[i][j:j+3] + grid[i+1][j:j+3] + grid[i+2][j:j+3]):
                return False
    return True

print(sudoku2(g))